stdlib/Profile/src/Profile.jl

```diff
@@ -505,15 +503,26 @@ By default metadata such as threadid and taskid will be stripped. Set `include_m
 """
 function fetch(;include_meta = false)
     maxlen = maxlen_data()
+    data = Vector{UInt}(undef, maxlen)
+    ccall(:jl_profile_lock, Cvoid, ())
+    # N.B. GC allocations before unlock might cause a deadlock
     len = len_data()
-    if is_buffer_full()
-        @warn """The profile data buffer is full; profiling probably terminated
-                 before your program finished. To profile for longer runs, call
-                 `Profile.init()` with a larger buffer and/or larger delay."""
+    GC.@preserve data begin
+        data_ptr = get_data_pointer()
+        tail = maxlen - len
+        unsafe_copyto!(pointer(data), data_ptr+sizeof(UInt)*len, tail)
+        unsafe_copyto!(pointer(data)+sizeof(UInt)*tail, data_ptr, len)
+    end
+    ccall(:jl_profile_unlock, Cvoid, ())
+    if data[1] == 0
```

Comment:
This is wrong: we need to make sure that we're not truncating a valid block end at the beginning of the buffer.